cmd/kube-controller-manager/app/controllermanager.go

```diff
@@ -162,7 +162,9 @@ func Run(s *options.CMServer) error {
 	}
 
 	if !s.LeaderElection.LeaderElect {
-		run(nil)
+		stopCh := make(chan struct{})
+		defer close(stopCh)
+		run(stopCh)
```

Comment:
I'm confused how this changes any behavior. The `defer close(stopCh)` should never run, since run shouldn't return until stopCh is closed. Receiving from a nil channel blocks forever, which would behave the same as receiving from an unclosed stopCh. Was something in run() explicitly checking if the passed channel was nil?

Reply:
With stopCh = nil, the graph_builder StartMonitors function exits without starting the monitors and no garbage collection happens.  See https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/controller/garbagecollector/graph_builder.go#L273

I don't know if creating a dummy stopCh is the best solution, but it let things follow the same path that is taken with leader-elect=true, which seems to be better tested.

This is my only foray into go, but it does seem unlikely that the deferred close serves any purpose.

Reply:
@liggitt See my response above.

Reply:
ah, then it seems better to fix the garbage collector specific assumptions in https://github.com/kubernetes/kubernetes/blob/release-1.8/pkg/controller/garbagecollector/graph_builder.go#L85 by setting up a non-nil dummy stopCh in GraphBuilder#Run if a nil stopCh is passed in. otherwise we still have the possibility of other callers making the same mistake

Reply:
@liggitt  If there was some other caller of GraphBuilder#Run, is it safe to assume that we should create a dummy stop channel if one is not provided?  I think it would be safer to require / expect that a stop channel be provided and maybe GraphBuilder#Run should fail in some fashion if one is not provided.  Other comments in graph_builder suggest that a stop channel is expected.

Maybe GarbageCollector#Run (https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/garbagecollector.go#L129) or startGarbageCollectorController (https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/core.go#L344) would be better places.

Hopefully other controllers that use the stop channel do not need it to be set.

And not seeing a better communication channel... I am getting in well past my comfort level with go and internals of Kubernetes.  If this is not the right way or place to fix this, I would prefer to leave this to others.

Reply:
Thank you!